cpp/src/strings/utilities.cuh

```diff
@@ -108,6 +111,84 @@ auto make_strings_children(
+ *        chars memory.
+ * @param strings_count Number of strings.
+ * @param mr Device memory resource used to allocate the returned columns' device memory.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @return offsets child column and chars child column for a strings column.
+ */
+template <typename SizeAndExecuteFunction>
+std::tuple<std::unique_ptr<column>, std::unique_ptr<column>, rmm::device_buffer, size_type>
+make_strings_children_with_null_mask(
+  SizeAndExecuteFunction size_and_exec_fn,
+  size_type strings_count,
+  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, strings_count + 1, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view          = offsets_column->mutable_view();
+  auto d_offsets             = offsets_view.template data<int32_t>();
+  size_and_exec_fn.d_offsets = d_offsets;
+
+  // This is called twice: once for offsets and once for chars
+  auto for_each_fn = [strings_count, stream](SizeAndExecuteFunction& size_and_exec_fn) {
+    thrust::for_each_n(rmm::exec_policy(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       strings_count,
+                       size_and_exec_fn);
+  };
+
+  // Compute the string sizes, storing in `d_offsets` (negative output sizes mean null strings)
+  for_each_fn(size_and_exec_fn);
+
+  // Use the string sizes to compute null_mask and null_count of the output strings column
+  auto [null_mask, null_count] = cudf::detail::valid_if(
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    [d_offsets] __device__(auto const idx) { return d_offsets[idx] >= 0; },
+    stream,
+    mr);
+
+  // Compute the offsets
+  auto const iter_trans_begin = thrust::make_transform_iterator(
+    d_offsets, [] __device__(auto const size) { return size < 0 ? 0 : size; });
+  thrust::exclusive_scan(
+    rmm::exec_policy(stream), iter_trans_begin, iter_trans_begin + strings_count + 1, d_offsets);
+
+  // Now build the chars column
+  auto const bytes  = cudf::detail::get_value<int32_t>(offsets_view, strings_count, stream);
+  auto chars_column = create_chars_child_column(strings_count, bytes, stream, mr);
+
+  // If all the strings are empty or null, `bytes` will be `0` thus the `d_chars` pointer will has
+  // `nullptr` value. Thus, we need to set an arbitrary pointer value to `d_chars` to prevent the
+  // string sizes to be computed again. It is safe to do so, because in this case the strings column
+  // has all empty or all null string elements so nothing will be copied onto `d_chars`.
+  size_and_exec_fn.d_chars =
+    bytes > 0 ? chars_column->mutable_view().template data<char>() : reinterpret_cast<char*>(0x1);
```

Comment:
This seems very suspect. It could be true today, but this is the type of thing that can change and present issues.

Reply:
It seems like we want  `for_each_fn` will do nothing if `bytes==0` here?
If so why not just do this instead:
```
if( bytes > 0 ) { 
  for_each_fn(size_and_exec_fn);
}
```

Reply:
Oh yeah, why not!!!

Reply:
Done.